### PR TITLE
Fix critical unhandled error

### PR DESCRIPTION
**Pull request by lucagiovagnoli**
After having had an interest read from [Twisted Programming Essentials](http://www.amazon.com/Twisted-Network-Programming-Essentials-McKellar-ebook/dp/B00BT0IEJE/ref=sr_1_1?ie=UTF8&qid=1460649331&sr=8-1&keywords=twisted+programming+essentials) I got an idea about what could be the cause for issue #27.

**Possible causes**
1. First, the way we are catching exceptions might be problematic. [Here](https://github.com/Yelp/fido/blob/master/fido/fido.py#L81) we should at least return the failure or raise the exception, but we are only setting the future's error. The book states: 
   
   > If a callback or errback at level N doesn’t raise an Exception or return a Failure, control is passed to the callback at level N + 1. Note that this applies to errbacks! If an errback doesn’t produce an error, control passes to the callback chain. Control will criss-cross between the errback and callback chains depending on the results of processing the event."
2. Secondly, [this](https://github.com/Yelp/fido/pull/18/commits/488149342a512027bbf84ea7ad4088cd5b4faad4) is stated as a "please don't do" in [crochet documentation](https://crochet.readthedocs.org/en/1.4.0/api.html#run-in-reactor-asynchronous-results) as Twisted APIs are not thread safe and should not be called if not in the Twisted reactor thread. I believe we should put back the `@run_in_reactor` decorator at all costs.
   
   > Twisted’s APIs are not thread-safe, and so they cannot be called directly from another thread. Moreover, results may not be available immediately. The easiest way to deal with these issues is to decorate a function that calls Twisted APIs with crochet.wait_for. [...] wait_for is implemented using run_in_reactor, a more sophisticated and lower-level API.

**Solution**

I am proposing here a way simpler solution (literally by the [book](http://www.amazon.com/Twisted-Network-Programming-Essentials-McKellar-ebook/dp/B00BT0IEJE/ref=sr_1_1?ie=UTF8&qid=1460649331&sr=8-1&keywords=twisted+programming+essentials), chapter "Web Clients - Agent") which solves the issue of CRITICAL errors #27 and seems to work properly with our internal system. The general idea is that there is no need for futures as the EventualResult object from crochet already acts as some sort of future. Hence there's no need to set results/exceptions in the future ourselves, as the EventualResult does it already flawlessly and without risk of losing 'unhandled exceptions' as it was happening. 

**Testing**

Finally, I tested this client with our services and our acceptance suite was all green.

### CODE REVIEW

File: fido/fido.py [hunk 1]
```diff
@@ -24,6 +23,40 @@
...
+    bodyProducer = FileBodyProducer(StringIO(body))
+    # content-length needs to be removed because it was computed based on
+    # body but body is now being processed by twisted FileBodyProducer
+    # causing content-length to lose meaning and break the client.
+    # FileBodyProducer will take care of re-computing length and re-adding
+    # a new content-length header later.
+    headers = dict(
+        (key, value)
+        for (key, value) in headers.iteritems()
+        if key != 'Content-Length' and key != 'content-length'
```
Comment (NicolasEhrhardt) [inline fido/fido.py]: key.lower() != "content-length" is more readable.

Comment (lucagiovagnoli) [inline fido/fido.py]: oops, yep, I had just introduced the .lower() in PR #26.
This is a merge with a past branch of mine ended poorly.

Thanks for taking the time to review Nicolas!

File: fido/fido.py [hunk 2]
```diff
@@ -110,21 +139,12 @@ def finished_errorback(error):
...
-    deferred.addErrback(finished.errback)
-
-    if timeout is not None:
-        # Cancel the request if we hit the timeout
-        def cancel_timer(response):
-            if timer.active():
-                timer.cancel()
-            return response
-        timer = reactor.callLater(timeout, deferred.cancel)
-        finished.addBoth(cancel_timer)
```
Comment (lucagiovagnoli) [inline fido/fido.py:126]: just noticed I forgot to implement the timeout. 
will look for a solution to this.

File: fido/fido.py [hunk 3]
```diff
@@ -5,7 +5,6 @@
 import os
 from urlparse import urlparse
 
-import concurrent.futures
 import crochet
 from twisted.internet import reactor
```
Comment (laucia) [inline fido/fido.py:8]: importing reactor is something we know can lead to problems because it leads to https://github.com/itamarst/crochet/issues/45 like issues

File: fido/fido.py [hunk 4]
```diff
@@ -20,9 +19,46 @@
 
 
 DEFAULT_USER_AGENT = 'Fido/%s' % __about__.__version__
-
 DEFAULT_CONTENT_TYPE = 'application/json'
 
+# infinite timeouts are bad
```
Comment (laucia) [inline fido/fido.py]: a more descriptive explanation here ? (like crochet defaulting to infinite timeout yadi yada ...)

File: fido/fido.py [hunk 5]
```diff
@@ -70,37 +120,29 @@ def connectionLost(self, reason):
...
+@crochet.run_in_reactor
+def fetch_inner(url, method, headers, body, timeout, connect_timeout):
+    """
+    This function must be run in the reactor thread because it is calling
+    twisted API which is not thread safe.
+    See https://crochet.readthedocs.org/en/1.4.0/api.html#run-in-reactor-\
+    asynchronous-results for additional information.
+
+    The implementation of this function with regards to how to use the Twisted
+    Agent is following the official book Twisted Network Programming Essentials
```
Comment (laucia) [inline fido/fido.py:161]: page/chapter ?

File: fido/fido.py [hunk 6]
```diff
@@ -110,21 +152,45 @@ def finished_errorback(error):
...
 
-    if timeout is not None:
-        # Cancel the request if we hit the timeout
-        def cancel_timer(response):
-            if timer.active():
-                timer.cancel()
-            return response
-        timer = reactor.callLater(timeout, deferred.cancel)
-        finished.addBoth(cancel_timer)
+    # erroback which handles various types of twisted timeout errors
```
Comment (laucia) [inline fido/fido.py]: You should probably add a docstring that state that we could let these errors bubble up, but for sanity of usage, we are raising more user friendly crochet.TimeoutError exception to hide the twisted ones.

File: fido/fido.py [hunk 7]
```diff
@@ -110,21 +152,45 @@ def finished_errorback(error):
...
+            raise crochet.TimeoutError(
+                "Connection was closed by Twisted Agent because the HTTP "
+                "connection took to long to establish."
+            )
+        return error
+
+    deferred.addErrback(handle_timeout_errors)
+
+    # if the timeout is hit, set a timer to cancel the request
+    timer = reactor.callLater(timeout, deferred.cancel)
```
Comment (laucia) [inline fido/fido.py]: timer => cancel_timer_deferred ?

File: tests/fido_test.py [hunk 8]
```diff
@@ -244,14 +241,56 @@ def test_get_agent_with_http_proxy():
...
-        future.result()
+    with mock.patch(
+        'fido.fido.crochet.setup',
+        side_effect=ValueError('I failed :(')
+    ):
+        with pytest.raises(ValueError) as e:
+            fido.fido.fetch('http://some_url')
 
-    assert e.value.message == 'I failed :('
+        assert e.value.message == 'I failed :('
```
Comment (askedrelic) [inline tests/fido_test.py]: Could probably pull all of these 'I failed :(' into an ERROR_MESSAGE constant

File: fido/fido.py [hunk 9]
```diff
@@ -151,9 +217,11 @@ def get_agent(reactor, connect_timeout=None):
...
-          content_type=DEFAULT_CONTENT_TYPE, user_agent=DEFAULT_USER_AGENT,
-          headers=None, body=''):
+def fetch(
+    url, method='GET', headers=None, body='', timeout=DEFAULT_TIMEOUT,
+    connect_timeout=DEFAULT_CONNECT_TIMEOUT,
+    content_type=DEFAULT_CONTENT_TYPE, user_agent=DEFAULT_USER_AGENT,
+):
     """Make an HTTP request.
 
     :param url: the URL to fetch.
```
Comment (askedrelic) [inline fido/fido.py:264]: I can't comment on it, but :param timeout: docstring is inaccurate. There is a default now. Want to update the docstring?

File: fido/fido.py [hunk 10]
```diff
@@ -56,6 +92,20 @@ def dataReceived(self, data):
         self.buffer.write(data)
 
     def connectionLost(self, reason):
+        """
+        :param reason:
```
Comment (askedrelic) [inline fido/fido.py:100]: This might be out of scope/not useful, but is there a handy web link to the Twisted docs for why we are checking for these specific reasons? Are these the only reasons to check? Are we not checking some reasons?

This is dense code and pointing to more docs could be helpful.

**Comment by askedrelic** [comment]
+1 for style and docs. The why behind the change is rough to understand, but I think we will be in a better place with this.

**Comment by lucagiovagnoli** [comment]
merged with #29, fixed issues, improved tests and documentation. acceptance tested with our internal  ecosystem.

File: fido/fido.py [hunk 11]
```diff
@@ -4,26 +4,61 @@
...
 
 from . import __about__
 from .common import listify_headers
 
 
 DEFAULT_USER_AGENT = 'Fido/%s' % __about__.__version__
-
 DEFAULT_CONTENT_TYPE = 'application/json'
 
+# crochet defaults to infinite timeouts and that could cause hanging forever
```
Comment (askedrelic) [inline fido/fido.py]: How about "Sane defaults to prevent infinite timeouts"

File: fido/fido.py [hunk 12]
```diff
@@ -71,37 +122,32 @@ def connectionLost(self, reason):
...
+        If the server takes forever to send the response and if `timeout` is
+        not provided, we could block forever.
+    :param timeout: maximum time for the server to send a response back before
+        we abort.
+
+    :return: a crochet EventualResult wrapping a
+        twisted.internet.defer.Deferred object
+    """
+
+    from twisted.internet import reactor
```
Comment (askedrelic) [inline fido/fido.py]: You might want to add a few lines lines here explaining this import; the docstring explains the function and general handling but why are we importing reactor right here and not at the top of the file?

**Comment by laucia** [comment]
lg2m

File: fido/fido.py (hunk 12 above)
Comment (lucagiovagnoli) [inline fido/fido.py]: @laucia  and @askedrelic 
Let's do so, I will tackle this properly by adding good tests and documentation in the context of issue https://github.com/Yelp/bravado/issues/179

tl;dr will tackle this in next PR

**Comment by askedrelic** [comment]
🚢

**Comment by laucia** [comment]
```
       _~
    _~ )_)_~
    )_))_))_)
    _!__!__!_
    \______t/
```

```
```

File: fido/fido.py (hunk 11 above)
Comment (lucagiovagnoli) [inline fido/fido.py]: missed this, done now

**Comment by lucagiovagnoli** [comment]
amended - just fixing comments and typos

**Comment by prat0318** [comment]
will go through it over the weekend. but i think @jnb and @analogue will have better context.

**Comment by laucia** [comment]
When travis complete, I think we are good to go!
